text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,106 @@
+# Meta
+[meta]: #meta
+- Name: BOM inclusion in layer content metadata
+- Start Date: 2021-04-01
+- Author(s): [@samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes the inclusion of a `BOM` (Bill-of-materials) table in layer content metadata TOML files (`<layer-name.toml>`). This would allow the `BOM`s generated from a layer to have the same lifetime as the layer itself and avoid various gotchas that happen currently as a result of the BOM being defined in the `launch.toml` or `build.toml`.
+
+# Motivation
+[motivation]: #motivation
+
+## Why should we do this?
+
+Currently `BOM` is a buildpack level entity and has to be defined in either the `launch.toml` or the `build.toml`. This presents several limitations which may be alleviated the this proposal.
+
+Often times `BOM` describes the artifacts present in the layer. Having this present in the `launch.toml` or `build.toml` which are ephemeral between builds means that the `BOM` has to be generated for each build, even if the layer itself has not changed and is cached. 
+
+This leads to surprising behaviors for users at times when they may be generating a `BOM` as part of the layer generation process and may find that their `BOM` is missing if the `BOM` generation process is not called when the layer is assumed to be cached.
+
+Having the `BOM` be a part of the layer content metadata also means that we can pin-point which layer contributed a BOM entry without having to implement extra logic in downstream buildpacks.
+
+- What is the expected outcome?
+
+The lifecycle is able to generate the final `BOM` by combining the `BOM` entries from top level metadata files like `launch.toml` and `build.toml` and also the ones from `<layer-name>.toml` which handling layer caching correctly i.e. if the layer is cached and re-used in the new build - the `BOM` associated with that layer will also be re-used.
```

Comment:
Would it make more sense to remove `bom` entirely from `launch.toml` and `build.toml`? What would be the advantage in having it defined in more than one place?

Reply:
I guess one reason to have the buildpack bom is so that bom that is not associated with a layer still has a place to stay. For eg app related bom?

Reply:
I see. I am just curious, what would be an example of a BOM entry that isn't tied to a layer?

Reply:
Something like a C++ or go buildpack that builds the binary in the app workspace. Any build related BOM for the compiled binary won't be in a layer but would come from the app workspace. Would be weird to create a layer just to add a BOM.

Reply:
Could we still remove the `bom` from `build.toml`?